Common/cpp/Tools/BackgroundQueue.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <jsi/jsi.h>
+
+#include "WorkletRuntime.h"
+
+#include <atomic>
+#include <condition_variable>
+#include <memory>
+#include <queue>
+#include <string>
+#include <thread>
+#include <utility>
+#include <vector>
+
+using namespace facebook;
+using namespace react;
+
+namespace reanimated {
+
+class BackgroundQueue : public jsi::HostObject {
+ public:
+  explicit BackgroundQueue(const std::string &name);
+
+  ~BackgroundQueue() override;
+
+  std::string toString() const {
+    return "[BackgroundQueue \"" + name_ + "\"]";
+  }
+
+  jsi::Value get(jsi::Runtime &rt, const jsi::PropNameID &propName) override;
+
+  std::vector<jsi::PropNameID> getPropertyNames(jsi::Runtime &rt) override;
+
+  void push(
+      const std::shared_ptr<WorkletRuntime> &runtime,
+      const std::shared_ptr<ShareableWorklet> &worklet);
+
+ private:
+  void runLoop();
+
+  const std::string name_;
+  std::thread thread_;
+  std::atomic_bool running_{true};
+  std::mutex mutex_;
+  std::condition_variable cv_;
+  std::queue<std::pair<
+      std::shared_ptr<WorkletRuntime>,
+      std::shared_ptr<ShareableWorklet>>>
+      queue_;
```

Comment:
I wonder, do we actually want to have a multi-thread queue or rather to tightly couple queues to their designated threads? What's the philosophy behind it?

Reply:
At this point there's no philosophy, we simply want to allow users to run worklets in background. That being said, we don't want to create a new thread per each job, so we want to re-use existing threads. Since threads (and runtimes) can do at most one thing at once, we need to queue these operations. Currently, a `BackgroundQueue` represents a single background thread which has its own queue. There's no concept of thread pool or executors, these things will come later on.

Reply:
So is it necessary at the moment to store pairs here? Why wouldn't `WorkletRuntime` just be a class private field?

Reply:
Currently there's no relationship between thread/queue and worklet runtime, you can submit any worklet runtime to any background queue. The API is like this: `runOnBackgroundQueue(queue, runtime, worklet)`. We could pass the worklet runtime to `createWorkletRuntime` but then we bind runtime to the queue which is less flexible.

Reply:
Thanks, it makes more sense now!